src/account/PluginManager2.sol

```diff
@@ -0,0 +1,67 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.25;
+
+import {EnumerableSet} from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+
+import {IPlugin} from "../interfaces/IPlugin.sol";
+import {FunctionReference} from "../interfaces/IPluginManager.sol";
+import {FunctionReferenceLib} from "../helpers/FunctionReferenceLib.sol";
+import {AccountStorage, getAccountStorage, toSetValue} from "./AccountStorage.sol";
+
+abstract contract PluginManager2 {
+    using EnumerableSet for EnumerableSet.Bytes32Set;
+
+    error DefaultValidationAlreadySet(address plugin, uint8 functionId);
+    error ValidationAlreadySet(bytes4 selector, address plugin, uint8 functionId);
+    error ValidationNotSet(bytes4 selector, address plugin, uint8 functionId);
+
+    function _installValidation(
+        address plugin,
+        uint8 functionId,
+        bool shared,
+        bytes4[] memory selectors,
+        bytes calldata installData
+    ) internal {
+        FunctionReference validationFunction = FunctionReferenceLib.pack(plugin, functionId);
+
+        AccountStorage storage _storage = getAccountStorage();
+
+        if (shared) {
+            if (!_storage.defaultValidations.add(toSetValue(validationFunction))) {
+                revert DefaultValidationAlreadySet(plugin, functionId);
+            }
+        }
+
+        for (uint256 i = 0; i < selectors.length; ++i) {
+            bytes4 selector = selectors[i];
+            if (!_storage.selectorData[selector].validations.add(toSetValue(validationFunction))) {
+                revert ValidationAlreadySet(selector, plugin, functionId);
+            }
+        }
+
+        IPlugin(plugin).onInstall(installData);
+    }
+
+    function _uninstallValidation(
```

Comment:
thoughts on adding a post uninstall check, require(defaultValidation.length > 0), to guard against account bricking?

Reply:
While it may be useful, I don't think we need to add it to the spec and/or reference implementation itself, and can leave it as an implementation choice.